File: programmers/prob49191.java
package programmers;

public class prob49191 {
    static int N;
    static boolean[][] map;

    public static void main(String[] args) {
        // 2
        System.out.println(solution(5, new int[][] { { 4, 3 }, { 4, 2 }, { 3, 2 }, { 1, 2 }, { 2, 5 } }));
    }

    public static int solution(int n, int[][] results) {
        input(n, results);
        floydWashall();
        return getStudentAmount();
    }

    static int getStudentAmount() {
        int amount = 0;
        for (int i = 0; i < N; i++) {
            int count = 0;
            for (int j = 0; j < N; j++) {
                if (i == j)
                    continue;
                if (!map[i][j] && !map[j][i]) {
                    count++;
                }
            }
            if (count < 1)
                amount++;
        }
        return amount;
    }

    static void floydWashall() {
        for (int k = 0; k < N; k++) {
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if (map[i][k] && map[k][j])
                        map[i][j] = true;
                }
            }
        }
    }

    static void input(int n, int[][] results) {
        N = n;
        map = new boolean[N][N];
        for (int i = 0; i < results.length; i++) {
            map[results[i][0] - 1][results[i][1] - 1] = true;
        }
    }

}

File: programmers/prob49191.py
def floyd_washall(n, graph):
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if not graph[i][j]:
                    graph[i][j] = graph[i][k] and graph[k][j]
    for i in range(n):
        graph[i][i] = True
    return graph


def solution(n, results):
    map = [[False]*n for _ in range(n)]
    for winner, loser in results:
        map[winner-1][loser-1] = True

    map = floyd_washall(n, map)
    count = 0
    for i in range(n):
        is_ok = True
        for j in range(n):
            if i == j:
                continue
            if map[i][j] or map[j][i]:
                continue
            is_ok = False
            break
        count = count+1 if is_ok else count
    return count


print(solution(5, [[4, 3], [4, 2], [3, 2], [1, 2], [2, 5]]))
